docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+
+As a cluster operator, I want to be able to provision both "unmanaged" and "managed" Kubernetes clusters from the same management cluster, so that I can support different requirements and use cases as needed whilst using a single operating model.
+
+#### Story 3
+
+As a Cluster API provider implementor, I want to be able to return the control plane endpoint created by the `ControlPlane Provider`, so that it fits naturally with how most of the native Managed Kubernetes implementations works.
+
+#### Story 4
+
+As a Cluster API provider developer, I want guidance on how to incorporate a managed Kubernetes service into my provider, so that its usage is compatible with Cluster API architecture/features and its usage is consistant with other providers.
+
+#### Story 5
+
+As a Cluster API provider developer, I want to enable the ClusterClass feature for a Managed Kubernetes service, so that users can take advantage of an improved UX with ClusterClass-based clusters.
+
+#### Story 6
+
+As a cluster operator, I want to use Cluster API to provision and manage the lifecycle of worker nodes that utilizes my cloud providers' managed instances (if they support them), so that I don't have to worry about the management of these instances.
+
+#### Story 7
+
+As a service provider I want to be able to offer Managed Kubernetes clusters by using CAPI referencing my own managed control plane implementation that satisfies Cluster API contracts.
+
+### Design
+
+Below we are documenting API changes and controllers changes required to omit the `<Infra>Cluster` entirely and to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint.
+
+#### Core Cluster API changes
+
+This proposal does not introduce any breaking changes for the existing "core" API. More specifically:
+
+The existing Cluster API types are already able to omit the `<Infra>Cluster`:
+
+- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
+- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.
+
+In order to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  hostname: 'name-1234567890.region.elb.amazonaws.com'
+  port: 1234
+  type: ExternalControlPlaneEndpoint
+```
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterEndpoint
+spec:
+  ip: 10.40.85.102
```

Comment:
For this API to be future proof are we planning to make ip and port slices?
Or should this be an slice of endpoints so in future each element of the slice can have a type?

Reply:
That's an interesting question... Maybe each endpoint should be its own object and have an owner ref to the control plane object?

Reply:
This should probably be ~close to what's in https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener 

A slice of ~Listeners (like in gateway networking) with a hostname and port should be sufficient to describe the endpoints?

Reply:
I think we want to only have one IP and port per ClusterEndpoint (singular endpoint).

Should we update the spec to include an ownerRef?

```
spec:
...
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AWSManagedControlPlane
    name: my-eks-cluster
    namespace: my-eks-cluster-ns
```

or

```
spec:
...
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureManagedControlPlane
    name: my-aks-cluster
    namespace: my-aks-cluster-ns
```

Reply:
Not sure if it's a good idea to add a ref in spec as ownerRef if Kubernetes has metadata.ownerRef for this purpose

Reply:
We can have it as a slice and limit to a single one via a webhook or openapi validation, that opens the possibility later to handle more than one

Reply:
What are we planning to make slice of? We can't make both host and port slices, as there are 1:1 relationships between the two of those properties. IMO any "set of endpoints" implementations can employ slices of this new type, rather than having a `ClusterEndpoint` (singular) represents sets of hosts and ports.

WDYT?

Reply:
A slice of `listeners`? https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener

Reply:
Thx. Can someone explain how "a singular endpoint containing multiple endpoints" is preferable to "multiple endpoints"? Why would an endpoint have multiple endpoints? (semantically speaking, a "listener" in this case is just a synonym for "endpoint")

Reply:
my two cents is to keep it simple because this covers the use case in the scope of the current work (the control plane endpoint) and unblock this work focused on managed K8s enablement.
Considering that this will land in 1.7, there are 4-5 months of time to do another iteration on this proposal/add another proposal extending this CR making it possible to support more control endpoints/more complex types of endpoints